test: Repro 'dune fmt' crash in presence of Nix result

Julow opened the pull request:
Reproduction case for a crash when using `dune fmt`.
Perhaps formatting shouldn't follow symlinks ? Also, `dune build` doesn't seem to have any problem with that, perhaps because the linked directory doesn't contain a dune file.
Perhaps symlinked directories should be considered `vendored_dirs` automatically ?

Julow committed:
test: Repro 'dune fmt' crash in presence of Nix result

A 'ocamlformat' executable is added to the test to avoid depending on
OCamlformat.

Signed-off-by: Jules Aguillon <jules@j3s.fr>

anmonteiro reviewed (commented):
I have seen this before and it can be a bit puzzling. 

IMO the issue here isn't that dune is following symlinks, but rather that the files in the nix store `/nix/store/...` have pretty locked-down permissions as one would expect.


I'm not saying that dune is perfect, just trying to point out the actual cause. There might be work to be done in dune itself.

Julow commented:
Is here a case where we want the files in the symlinked directory be formatted ?

Leonidas-from-XIV commented on test/blackbox-tests/test-cases/nix-result-in-tree.t/run.t:
```diff
@@ -0,0 +1,55 @@
+  $ chmod -R a-w "$RESULT"
+  $ ln -s "$RESULT" result
+
+  $ dune fmt --auto-promote
```
Isn't `--auto-promote` the default for `dune fmt`?

Leonidas-from-XIV commented on test/blackbox-tests/test-cases/nix-result-in-tree.t/dune:
```diff
@@ -0,0 +1,2 @@
+(executable
+ (name ocamlformat))
```
I don't think this has anything to do with Nix per-se, thus I would suggest renaming it to something akin to `read-only-symlink-target.t`.

Leonidas-from-XIV commented on test/blackbox-tests/test-cases/nix-result-in-tree.t/run.t:
```diff
@@ -0,0 +1,55 @@
+  Execution will pass over me and through me.  And when it has gone past, I
+  will unwind the stack along its path.  Where the cases are handled there will
+  be nothing.  Only I will remain.
+  [1]
```
Could you filter the output a bit so it is more reproducible? The locations in the stacktrace are probably going to change over time. A simple `grep EACCES` is probably sufficient to show there's a very specific error that happens.

Leonidas-from-XIV commented on test/blackbox-tests/test-cases/nix-result-in-tree.t/run.t:
```diff
@@ -0,0 +1,55 @@
+  be nothing.  Only I will remain.
+  [1]
+
+Allow Dune to remove temporary files (Dune crashes without this):
```
This is for the calling dune that is running `runtest` I assume?

Julow replied:
Yes, it seems that it fail at cleaning up the temporary files created during the test.

Leonidas-from-XIV reviewed (commented):
Thanks for the repro case; I've posted some comments.

As to how to deal with projects like this, if we can't find the proper solution I would suggest adding it to the dev-meeting agenda to discuss.

rgrinberg committed:
fix: improve error messages

Make user errors in hooks just regular errors

Simplify the erorr for promotion failures

Signed-off-by: Rudi Grinberg <me@rgrinberg.com>

rgrinberg commented:
Thanks. This is just required a little more code for a better error message.

Julow commented on test/blackbox-tests/test-cases/read-only-symlink-target.t/run.t:
```diff
@@ -0,0 +1,22 @@
+  _build/default/result/.formatted/foo.ml differ.
+  Promoting _build/default/result/.formatted/foo.ml to result/foo.ml.
+  Error: failed to promote result/foo.ml
+  Permission denied
```
Thanks for the fix! With this output, it's clearer that removing `result` would fix the issue.
But couldn't Dune avoid this issue ? Formatting files that are outside the workspace is surprising but could be part of someone's workflow. However, formatting files that are read-only feels like something Dune could avoid without bothering anyone.

rgrinberg replied:
> However, formatting files that are read-only feels like something Dune could avoid without bothering anyone.

That's an interesting suggestion. It does seem like an improvement for the current behavior. There's no need to setup rules if we know they're good to fail. Do you want to try implementing this?

Julow replied:
I've implemented that in https://github.com/ocaml/dune/pull/11490 but I have the feeling that is not the right solution. `result` links might interfere with other things than formatting and I think they should not be read by Dune.
